# String

* 字符串是 Python 中最常用的数据类型。我们可以使用引号( `'` 或 `"` )来创建字符串。
* 创建字符串很简单，只要为变量分配一个值即可。
* `'''` and `"""`：多行注释

In [3]:
# 三引号定义
name = """
黑马程序员
"""
print(name)
name


黑马程序员



'\n黑马程序员\n'

In [4]:
# 字符串内包含单/双引号
name1 = "'黑马程序员'"
name2 = '"和马程序员"'
print(name1)
print(name2)

'黑马程序员'
"和马程序员"


## 1. Python 访问字符串中的值

* `变量[头下标:尾下标:step]`
    * 索引值以 0 为开始值，-1 为从末尾的开始位置。

In [61]:
str1 = 'Runoob' # 6 bytes

* 索引有6个
    * 分别为 0 1 2 3 4 5
    * 分别为 -6 -5 -4 -3 -2 -1
* 截取有7个（6个字符），按照光标分隔
    * 分别为 0 1 2 3 4 5 6
    * 分别为 0 -1 -2 -3 -4 -5 -6
    * 类似于字符间的分隔
    * 简单记法：截取字符串中的一部分，遵循左闭右开原则，str[0:2] 是不包含第 3 个字符的

In [62]:
print(str1[0])
print(str1[-6])

print(str1[5])
print(str1[-1])

print(str1[0:])
print(str1[:6])

R
R
b
b
Runoob
Runoob


In [85]:
print(str1[::-2])
print(str1[1:2])

bou
u


In [82]:
print(str1[0:-4])
print(str1[-3:-1])

Ru
oo


## 2. 字符串operations

1. `+` 
2. `* `
3. `[ ]` 索引选取
4. `[:]` 截取
5. `in` 成员操作符
5. `not in`
6. `/r /R` raw string
7. `%` formate string

In [65]:
# + 和 * 
str2 = "stuAwesome"
str3 = "Hello World"
print(str3[:6]+str2)

print((str3[:6]+str2 + "\t")*3)

Hello stuAwesome
Hello stuAwesome	Hello stuAwesome	Hello stuAwesome	


## 3. 字符串转义
* 在需要在字符中使用特殊字符时，python 用反斜杠 \ 转义字符。\
* 转义字符标：（如下只有部分）

In [66]:
# \(在行尾时)
# 续行符
print("line1 \
line2 \
line3")

line1 line2 line3


In [67]:
# \\
# 反斜杠符号
print("\\")

\


In [68]:
# \
# 单引号
print('\'')

'


In [69]:
#\a
# 执行后电脑有响声
print("\a")




In [70]:
# \n 换行
print('Stu\nAwesome')

Stu
Awesome


In [71]:
# \t Tab 横向制表符
print('Stu\tAwesome')

Stu	Awesome


In [72]:
# \v 纵向制表符
print('Stu \v Awesome')

Stu  Awesome


In [73]:
# \r
# 将 \r 后面的内容移到字符串开头，并逐一替换开头部分的字符，直至将 \r 后面的内容完全替换完成。

# google被替换
print('google runoob taobao\r123456')

123456


In [74]:
# \f
# 换页
print("Hello \f World!")

Hello  World!


In [75]:
# \yyy 八进制数，y 代表 0~7 的字符，例如：\012 代表换行。
print("\110\145\154\154\157\40\127\157\162\154\144\41")

Hello World!


In [76]:
# \xyy 十六进制数，以 \x 开头，y 代表的字符，例如：\x0a 代表换行
print("\x48\x65\x6c\x6c\x6f\x20\x57\x6f\x72\x6c\x64\x21")

Hello World!


## 4.  Python字符串格式化

```
    占位符号描述
      %c	 格式化字符及其ASCII码
      %s	 格式化字符串
      %d	 格式化整数
      %u	 格式化无符号整型
      %o	 格式化无符号八进制数
      %x	 格式化无符号十六进制数
      %X	 格式化无符号十六进制数（大写）
      %f	 格式化浮点数字，可指定小数点后的精度
      %e	 用科学计数法格式化浮点数
      %E	 作用同%e，用科学计数法格式化浮点数
      %g	 %f和%e的简写
      %G	 %f 和 %E 的简写
      %p	 用十六进制数格式化变量的地址
格式化操作符辅助指令:

符号	功能
*	定义宽度或者小数点精度
-	用做左对齐
+	在正数前面显示加号( + )
<sp>	在正数前面显示空格
#	在八进制数前面显示零('0')，在十六进制前面显示'0x'或者'0X'(取决于用的是'x'还是'X')
0	显示的数字前面填充'0'而不是默认的空格
%	'%%'输出一个单一的'%'
(var)	映射变量(字典参数)
m.n.	m 是显示的最小总宽度,n 是小数点后的位数(如果可用的话)
Python2.6 开始，新增了一种格式化字符串的函数 str.format()，它增强了字符串格式化的功能。
```

In [6]:
num1 = 11
num2 = 11.345 
print(f"{num1:5}")
print(f"{num2:7}")

   11
 11.345


* m：控制宽度，要求是大于m宽度的数字
    * m = 数字+一位小数点
    * 7.2f 对 11.345 进行设置 结果是 [空格][空格][空格]11.35 七个宽度，两位小数
* .n：控制精度 (四舍五入 会失真)

### 可以format表达式

In [10]:
print(f"{1*1}")

1


### 方法一：傻瓜式
* string 与 变量间 有个`%`
* %s 占位为字符串 - **表达式**
* %d 整型
* %f

In [77]:
print("我今年%d岁了，存款只有%0.2f，我叫%s"%(20,1.23234,"Stu"))

我今年20岁了，存款只有1.23，我叫Stu


In [12]:
print("%s" % type(1))

<class 'int'>


### 方法二：`{ }`.format(变量1, ...)

In [78]:
print("我今年{0}岁了，存款只有{1:.2f}，我叫{2}".format(20,1.23234,"Stu"))
print("我今年{:>5}岁了，存款只有{:<8}，我叫{:^8}".format(20,1.23234,"Stu"))

我今年20岁了，存款只有1.23，我叫Stu
我今年   20岁了，存款只有1.23234 ，我叫  Stu   


### 方法三： fstring

* 只用这一种就可以

In [79]:
print(f"存款只有{1123.1231:.2f}，我叫{2}")

存款只有1123.12，我叫2


## 5. Unicode 字符串

1. 编码：字符串 -> 010101 byte 的规则
    * python里，每次导入导出数据string和bytes之间转换 - encode and decode，最好都使用UTF-8
2. 一开始只有英文编码，只能翻译数字和英文大小写字母：普通字符串是以8位ASCII码进行存储的
3. 中国有 GB232 和 GBK
4. 各国有各国很混乱
    * 统一：Unicode标准的UTF-8



* 在Python2中，普通字符串是以8位ASCII码进行存储的
* 而Unicode字符串则存储为16位unicode字符串，这样能够表示更多的字符集。使用的语法是在字符串前面加上前缀 u。
* 在Python3中，所有的字符串都是Unicode字符串。

##  6. string 的 built-in function
* 共有40个，只展示部分
* 用到去查去记

## 6.1 method 方法 和 function 函数

* method是特殊函数，不能独立使用，只能针对某个对象object使用
* method只针对某个对象使用，因为有些参数被隐含地传过去
* OOP 面向对象编程

In [88]:
# method
'abcd'.upper()

'ABCD'

In [91]:
# upper函数 不在 python 宇宙中
# len() 在
a = 'abcd'
len(a)

upper(a)

NameError: name 'upper' is not defined

### string的methond

* l：从左搜索
* r：从右搜索

1. 大小写转换
```python
str.lower()
str.upper()
str.capitalize()
str.title() #返回"标题化"的字符串,就是说所有单词都是以大写开始，其余字母均为小写(
str.swapcase()
```

2. 搜索
```python
str.count()
str.startswith()
str.endswith()
str.find()
str.rfind()
str.index()
str.rindex()
str.replace()
```
3. 填充 or 剔除
```python
str.center()
str.ljust()
str.zfill()
str.strip()
str.lstrip()
str.rstrip()
```
4. 分割 合并
```python
str.partition()
str.rpartition()
str.split()
str.rsplit()
str.splitlines()
str.join()
```
5. 判断
```python
str.islower()
str.isupper()
str.istitle()
str.isdigit()
str.isdecimal()
str.isnumeric()
str.isalpha()
str.isalnum()
str.isprintable()
str.identifier()
```

1. 大小写转换
```python
str.lower()
str.upper()
str.capitalize()
str.title() #返回"标题化"的字符串,就是说所有单词都是以大写开始，其余字母均为小写(
str.swapcase()
```

In [100]:
text = "asdjHK DJA SHDkjl hasdha sHJKkjf"

In [97]:
text.lower()

'asdjhkdjashdkjlhasdhashjkkjf'

In [98]:
text.upper()

'ASDJHKDJASHDKJLHASDHASHJKKJF'

In [101]:
# 句中单词首字母大写 第一次单词大写
text.capitalize()

'Asdjhk dja shdkjl hasdha shjkkjf'

In [103]:
# 句中每个单词首字母大写
text.title()

'Asdjhk Dja Shdkjl Hasdha Shjkkjf'

In [104]:
#大小写转换
text.swapcase()

'ASDJhk dja shdKJL HASDHA ShjkKJF'

2. 搜索
```python
str.count()
str.startswith()
str.endswith()
str.find()
str.rfind()
str.index()
str.rindex()
str.replace()
```

In [105]:
text.count("s")

3

In [109]:
# 加上数字，代表从哪里开始
text.count("s",2)

2

In [110]:
# 加上数字，代表从哪里开始 结束
text.count("s",1,3)

1

In [111]:
company1 = "深圳市派森"
company2 = '成都市派森'
company1.startswith("深圳")

True

In [112]:
company2.startswith("深圳")

False

In [113]:
text.find("s")

1

In [115]:
text.rfind("s")

25

In [117]:
text.rfind("m")
# -1 代表没有这个字母

-1

In [151]:
text1 = text.replace("s","\n")
print(text1)

a
djHK DJA SHDkjl ha
dha 
HJKkjf


In [149]:
# 决定替换几个单词
text.replace("s","m",1)

'amdjHK DJA SHDkjl hasdha sHJKkjf'

3. 填充 or 剔除
* 在str前面或后面填充内容，第一个参数数字代表了这个str的最后总长度
```python
str.center()
str.ljust()
str.zfill()
str.strip()
str.lstrip()
str.rstrip()
```

In [144]:
"2023".rjust(12,'Y')

'YYYYYYYY2023'

In [130]:
"2023".zfill(9)

'000002023'

In [133]:
#两侧填充
text.center(50,"#")

'#########asdjHK DJA SHDkjl hasdha sHJKkjf#########'

In [135]:
# 去除做两边制定字段、
(text.center(50,"#")).strip("#")

'asdjHK DJA SHDkjl hasdha sHJKkjf'

In [137]:
# 去除做两边制定字段、
#(text.center(50,"#")).strip("#")

#去除两边空格
#两侧填充
text1 = text.center(50," ")
text1.strip()

'asdjHK DJA SHDkjl hasdha sHJKkjf'

4. 分割 合并
* eg. 把str拆分成list

```python
str.partition()
str.rpartition()
str.split()
str.rsplit()
str.splitlines()
str.join()
```

In [138]:
text.split("s")

['a', 'djHK DJA SHDkjl ha', 'dha ', 'HJKkjf']

In [159]:
# 按空格分割 
# stu 没有成功实现
text.splitlines()

['asdjHK DJA SHDkjl hasdha sHJKkjf']

In [161]:
#每个字符前插入前面的str
"$".join(text)

'a$s$d$j$H$K$ $D$J$A$ $S$H$D$k$j$l$ $h$a$s$d$h$a$ $s$H$J$K$k$j$f'

In [162]:
#每个字符前插入前面的str
text.join("分割")

'分asdjHK DJA SHDkjl hasdha sHJKkjf割'

5. 判断
* 有多种方法
```python
str.islower()
str.isupper()
str.istitle()
str.isdigit()
str.isdecimal()
str.isnumeric()
str.isalpha()
str.isalnum()
str.isprintable()
str.identifier()
```

In [163]:
"1".isdigit()

True

In [164]:
help(str.isdigit)

Help on method_descriptor:

isdigit(self, /)
    Return True if the string is a digit string, False otherwise.
    
    A string is a digit string if all characters in the string are digits and there
    is at least one character in the string.



In [166]:
# 查看 isdigit 的 doc文档
print(str.isdigit.__doc__)

Return True if the string is a digit string, False otherwise.

A string is a digit string if all characters in the string are digits and there
is at least one character in the string.
